### ISSUE TITLE
All main options in menubar "View" doesn't work

### ISSUE THREAD

**Issue by simonbcn** [transmission-remote-gtk/transmission-remote-gtk#25]
Arch Linux 64 bits
transmission-remote-gtk git version (_compiled today_)
GNOME Shell 3.20.1

When I try to show/hide some option (_State selector_, _Directory filters_, _Tracker filters_, _Torrent Details_) or change the visual style (_Transmission style_, _Transmission compact style_, _Classic style_)  in menu bar "_View_", it does nothing. It ignores me completely. :cry:
I suppose this is a bug due to a new gtk3 change in last Gnome version.

In console it shows this error several times:

```
(transmission-remote-gtk:30595): Json-CRITICAL **: json_node_set_int: assertion '!node->immutable' failed
```

**Comment by leechunkin** [comment]
GTK changes json_parser_get_root() to return immutable node, but json_node_copy() does not really copy the node. Therefore the loaded prefs tree is immutable and every modification will fail. This bug affects all settings.

**Comment by TingPing** [comment]
I swear it worked for me recently and that change you mention is a year old.

**Comment by simonbcn** [comment]
I have downloaded and compiled this repository again and it still fails:
```
$ src/transmission-remote-gtk

(transmission-remote-gtk:16481): Gtk-WARNING **: Allocating size to TrgMainWindow 0x5616d02ea630 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(transmission-remote-gtk:16481): Gtk-WARNING **: Allocating size to TrgMainWindow 0x5616d02ea630 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(transmission-remote-gtk:16481): Gtk-WARNING **: Allocating size to TrgMainWindow 0x5616d02ea630 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-795050111" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-750998857" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Gtk-WARNING **: Allocating size to TrgMainWindow 0x5616d02ea630 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-1847733345" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-1235924857" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-216244044" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-402656965" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-1218511271" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-648732130" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-1762834127" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_boolean: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): GLib-GObject-WARNING **: value "-2131671645" of type 'gint64' is invalid or out of range for property 'eta' of type 'gint64'

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_int: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_int: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_int: assertion '!node->immutable' failed

(transmission-remote-gtk:16481): Json-CRITICAL **: json_node_set_int: assertion '!node->immutable' failed
```

IMHO, the problem is that this project was made for GTK2 and the adaptation to GTK3 is very incomplete: inconsistent icons, use the deprecated widgets/methods,...
Moreover, GTK3 is constantly evolving and that can cause certain _things_ that worked before, stop working.

This is a screen record: https://dl.dropboxusercontent.com/u/1466192/captura.webm

The only option that "_responds_" is the new `Directories first` but it doesn't work well. The submenu `Sort` and the option `Statistics` work.

The GTK version in use is `3.20.9-1`

**Comment by TingPing** [comment]
This is entirely about the json-glib usage and nothing to do with gtk.

**Comment by simonbcn** [comment]
json-glib 1.2.2-1

**Comment by timdp** [comment]
Same issue on Ubuntu 16.10. It didn't happen on 16.04 so the newer json-glib would have to have introduced a breaking change?

**Comment by geebs61** [comment]
Is there a workaround? This is the only mention of this error I can find online, and while I like Transmission-remote, it's worthless to me if I can't get it into classic style.

**Comment by timdp** [comment]
This is still happening on Ubuntu 17.04. What needs to happen to get it working again?

As @leechunkin seems to think it's related to json-glib suddenly returning immutable trees, how about cloning the immutable node into a mutable one and proceeding with that one?

What I still don't understand is how this issue would have originated. If I google `assertion '!node->immutable' failed`, I only find transmission-remote-gtk issues. That would indicate that the issue isn't widespread, so it seems to me like transmission-remote-gtk has always been the odd one out in its usage of json-glib. Perhaps the intention has always been to have an immutable tree and they just started enforcing it?

**Comment by TingPing** [comment]
> Perhaps the intention has always been to have an immutable tree and they just started enforcing it?

That would be my guess. Not really a solution to this but I've created a new client and don't work on this anymore: https://github.com/TingPing/transmission-remote-gnome

**Comment by timdp** [comment]
Interesting. I look forward to the first full-fledged release.

If you're not planning to fix this issue, maybe you should just close it?

**Comment by TingPing** [comment]
> If you're not planning to fix this issue, maybe you should just close it?

Nah, the bug still exists and people are free to fix it.

**Comment by sronsiek** [comment]
Observation:

When starting transmission-remote-gtk the first time, entering the server details & connecting causes a local .config/transmissions-remote-gtk/config.json to be written. In my case the size of this file was 1106 bytes.

At this stage, changing view options still works. You can test by renaming the cfg file white the prog is not running.

Once  transmission-remote-gtk is closed, the config file is again written, but this time there appears to be additional data in the file - the size is now 2270. 

On subsequent re-start, the view options no longer work ...

**Comment by alvenca** [comment]
This issue is solved with json-glib 1.2.8

**Comment by timdp** [comment]
@alvenca Interesting. Unfortunately, on Ubuntu in particular, `apt show libjson-glib-1.0.0` currently tells me it's version 1.2.2-1.

**Comment by alvenca** [comment]
@timdp Yeah, tested on Arch
I belive the actual commit was this https://git.gnome.org/browse/json-glib/commit/?h=json-glib-1-2&id=0cb1db37f8a1edf95f8d14c8868091804e54e4c3

**Comment by timdp** [comment]
Doesn't look like a huge change. I might try backporting it and compiling from source. Thanks for the heads up!

**Comment by timdp** [comment]
Confirming that backporting the [patch](https://git.gnome.org/browse/json-glib/commit/?h=json-glib-1-2&id=0cb1db37f8a1edf95f8d14c8868091804e54e4c3) reported by @alvenca fixes this issue on Ubuntu 17.04. Here's how I did it:

1. Make sure you have a `deb-src` entry for `zesty/main` in `/etc/apt/sources.list`. Additionally, run `sudo apt install build-essential devscripts` to make sure you're set up for building packages.
2. Create a temporary directory for the build and enter it: `mkdir json-glib && cd json-glib`
3. Grab the sources: `sudo apt update && sudo apt-get source json-glib`
4. Navigate to the sources: `cd json-glib-1.2.2`
5. Apply the [patch](https://git.gnome.org/browse/json-glib/patch/?id=0cb1db37f8a1edf95f8d14c8868091804e54e4c3) by modifying `json-glib/json-parser.c` using a text editor or running it through `patch`.
6. Grab the build dependencies: `sudo apt-get build-dep json-glib`
7. Update the changelog: `dch -i`
8. Build the packages: `debuild -b -us -uc` (this might take a while)
9. Navigate up and find the `.deb` files that were just built: `cd ..`
10. Install the packages: `sudo dpkg -i libjson-glib-1.0-common_1.2.2-1ubuntu1_all.deb libjson-glib-1.0-0_1.2.2-1ubuntu1_amd64.deb`

With those packages installed, transmission-remote-gtk is back to normal on my system.

Many thanks to @alvenca for pointing me in the right direction!
